Align to BIP78

Kukks opened the pull request:
closes #2

This removes the address format restriction, and adds the new optional parameters and checks associated with that.

Kukks committed:
Align to BIP78

Kukks committed:
redo it all

Kukks committed:
format

Kukks committed:
lint

Kukks committed:
lint format

junderw reviewed (changes requested):
Please fix the tests. CI should be passing, and ideally we should be upping the coverage threshold to 90 across the board.

Basically, by reading the test titles "describe x, it should y" we should be able to get an idea that all the points in the BIP are covered.

From a cursory glance, everything looks ok. But I can't merge this without tests passing at least.

Kukks committed:
More cleanup

Kukks committed:
tests

Overtorment commented:
@Kukks tests still fail

junderw committed:
Format and build

junderw commented:
After fixing the format problem and the build problem it seems as though tests are still failing.

junderw committed:
Set specific versions for devDeps

Overtorment commented:
attempt to use this branch in bluewallet throws  `Can not add duplicate data to input`

```
overtorment@laptop3 ~/Documents/BlueWallet (master) $ ./node_modules/.bin/jest     tests/unit/payjoin-transaction.test.js
 FAIL  tests/unit/payjoin-transaction.test.js
  PayjoinTransaction
    ✓ throws if smth is wrong with pj transaction (47ms)
    ✕ works (6ms)

  ● PayjoinTransaction › works

    Can not add duplicate data to input

      at Object.updateInput (node_modules/bitcoinjs-lib/node_modules/bip174/src/lib/utils.js:99:19)
      at Psbt.updateInput (node_modules/bitcoinjs-lib/node_modules/bip174/src/lib/psbt.js:46:13)
      at Psbt.updateInput (node_modules/bitcoinjs-lib/src/psbt.js:577:15)
      at PayjoinClient.run (node_modules/payjoin-client/src/client.js:165:23)
```

https://github.com/BlueWallet/BlueWallet/pull/1926

Kukks committed:
add 1 test and fix another

Kukks committed:
more tests and 1 more test fixed

Kukks committed:
format and other fixes in client

Kukks committed:
one more test!!!!!111shift+1!

Kukks committed:
unit testing like Ray Charles

Kukks committed:
format and build

junderw commented:
I have pushed a new image to `junderw/btcpay-client-test-server`

Uses latest NBXplorer and BTCPayServer tagged versions.

junderw commented:
Integration tests are failing because it's trying to fetch('bitcoin:xxx') and fetch needs an http/https link.

So the issue is in the parsing of the URL.

The URL parsing logic is very weird.

If the URL has one logic and the optional opts has another logic it looks as though it will overwrite the URL logic with the opts logic.

Since the URL represents what the server wants to do, if what WE want to do (opts) is incompatible it should throw an error.

Kukks commented:
> Integration tests are failing because it's trying to fetch('bitcoin:xxx') and fetch needs an http/https link.
> 
> So the issue is in the parsing of the URL.
> 
> The URL parsing logic is very weird.
> 
> If the URL has one logic and the optional opts has another logic it looks as though it will overwrite the URL logic with the opts logic.
> 
> Since the URL represents what the server wants to do, if what WE want to do (opts) is incompatible it should throw an error.

Yeah, the situation can be a little confusing:
The BIP21 payload can signal various things for BIP78 payjoin, which would set some of the optional parameters when communicating with the payjoin server. IMO, the payjoin client should be able to override these if they really wanted though.



> I have pushed a new image to `junderw/btcpay-client-test-server`
> 
> Uses latest NBXplorer and BTCPayServer tagged versions.

Thanks for the bump, will start fixing this part next. I fixed the rest of the spec tests and added a few more, but I may need to add more since coverage may be a bit down now.

junderw commented:
>  the payjoin client should be able to override these if they really wanted though.

If this is allowed it should be in the BIP. These kind of implementation details of "oh, the server requires you follow X but the client implementation allows people to override X which will cause a server error" are exactly what BIPs are designed to avoid.

Kukks committed:
remove unused funcs

Kukks committed:
use proper payjoin url not bip21

Kukks committed:
more coverage

Kukks committed:
f & b

Kukks commented:
> > the payjoin client should be able to override these if they really wanted though.
> 
> If this is allowed it should be in the BIP. These kind of implementation details of "oh, the server requires you follow X but the client implementation allows people to override X which will cause a server error" are exactly what BIPs are designed to avoid.

This was a misunderstanding. In my refactoring, I thought I had started requiring the full BIP21 instead of just the payjoin endpoint, but I was wrong. I've changed it now and the tests seem to be passing except for the 65% coverage quota (currently 60)

Kukks committed:
fix result

junderw committed:
Add html reporter for coverage

junderw commented:
There's quite a few branches missing.

I added html reporter to this branch, so run `npm run coverage` locally then open ./coverage/index.html in a web browser to get an idea of which branches are not covered...

It looks like a ton of the branches are major logic paths, and critical to test before release.

Ideally coverage should be >=90% across the board, but luckily it looks like most of the missing lines and statements are inside these branches, so if we get branches above 90% I think everything else will be over 90 too.

Kukks committed:
add guards for invalid payjoinclient opts

junderw commented:
Published 9241a826768b673f8d371f2d60db975d4139cccf as tag `beta` on npm. The version is `1.0.0-beta`

junderw commented on ts_src/client.ts:
```diff
@@ -1,55 +1,58 @@
+export interface PayjoinClientOptionalParameters {
+  disableOutputSubstitution?: boolean;
+  payjoinVersion?: number;
+  additionalfeeoutputindex?: number;
+  maxadditionalfeecontribution?: number;
```
camelCase

junderw commented on ts_src/utils.ts:
```diff
@@ -200,3 +93,58 @@ export function getInputIndex(
+  if (!payjoinParameters) {
+    return url;
+  }
+  const parsedURL = new URL(url);
```
Don't use URL.

junderw committed:
Use camelCase

junderw committed:
Use querystring

junderw commented:
Published 98f0e3b as tag `beta` on npm. The version is `1.0.0-beta2`

Overtorment commented:
querystring is also apparently very limited in RN. can `encodeURIComponent` be used instead?

Kukks committed:
Remove qs dependency

Overtorment commented:
just tested latest commit with bw, worked okay!

junderw commented:
Because we had to roll our own method for parsing querystrings, we need WAAAAAAAAAAY more tests. Test a bunch of edge cases. What if the server is malicious and tries to somehow break our wallet by putting in a bad URL with tons of ? or weird escaping loops (an escape of an escape of an escape of an escape symbol etc.)... String parsing is one of the most overlooked sources for vulnerabilities, and the URL is supplied by someone we are assuming is evil in other parts of the library (we check everything they tell us on our side, but URL is parsed blindly) We probably should throw some errors in certain cases of malformed URLs.

Kukks commented:
Ok ok, what if we revert to the original, and let BW provide a qs manager
override

On Wed, Oct 7, 2020 at 1:56 PM Jonathan Underwood <notifications@github.com>
wrote:

> Because we had to roll our own method for parsing querystrings, we need
> WAAAAAAAAAAY more tests. Test a bunch of edge cases. What if the server is
> malicious and tries to somehow break our wallet by putting in a bad URL
> with tons of ? or weird escaping loops (an escape of an escape of an escape
> of an escape symbol etc.)... String parsing is one of the most overlooked
> sources for vulnerabilities, and the URL is supplied by someone we are
> assuming is evil in other parts of the library (we check everything they
> tell us on our side, but URL is parsed blindly) We probably should throw
> some errors in certain cases of malformed URLs.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoinjs/payjoin-client/pull/3#issuecomment-704884879>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAN357T7LQHMCQOIHDLC4QDSJRJONANCNFSM4QDD4N2Q>
> .
>

junderw commented:
That would be fine.

Kukks committed:
Revert "Remove qs dependency"

This reverts commit 21830f6a6dedf9d2339cca598eb8b388c432a38b.

Kukks committed:
Revert "Use querystring"

This reverts commit 98f0e3b255223157d4bd277576d2df7dce3d2b30.

Kukks committed:
allow passing getEndpointUrl to client

Kukks committed:
allow overriding param writing in payjoin url

Overtorment commented:
I have just tested BW with lates payjoin-client from this PR, and everything worked well
![image](https://user-images.githubusercontent.com/1913337/95991565-d334da80-0e24-11eb-90ee-ef66d20563ea.png)

Kukks committed:
get proper test coverage!

junderw reviewed (approved):
LGTM.

We will publish under a major version bump.

junderw commented:
published to npm as v1.0.0

Overtorment commented:
so far so good

https://github.com/BlueWallet/BlueWallet/pull/2203


![image](https://user-images.githubusercontent.com/1913337/100120424-d41f4a00-2e6f-11eb-9346-c8ee18c59fc5.png)

![image](https://user-images.githubusercontent.com/1913337/100120456-daadc180-2e6f-11eb-9b99-fa4422a2587e.png)

junderw commented:
please also try some of the newer payjoin features if you can.

Overtorment commented:
how do I do that? I know only one payjoin-enabled server https://btc.donate.kukks.org
